### PR TITLE
Bluetooth: tester: Add support for BTP_GAP_SET_POWERED command

### LINKED ISSUE #67346: Enable bt-tester BTP_GAP_SET_POWERED cmd to call bt_disable() to power off controller
**Is your feature request related to a problem? Please describe.**
We're implementing a BLE FW reload feature on Zephyr bt-tester, which intends to shutdown BLE controller when not needed, and will reload and reinit BLE controller when needed. Currently Register GAP Service (00 03 ff 01 00 01)cmd can help to call `bt_enable`-->HCI `open `to load and init BLE controller, but there is no any cmd in bt-tester calling the `bt_disable `then HCI `close `to shutdown BLE controller       

**Describe the solution you'd like**
We found one standard BTP GAP cmd can be used to achieve shutdown BLE controller function, which could meet our requirement perfectly we think.  It's **Opcode 0x05 - Set Powered command/response,** which is used to power on/off a controller per BTP protocol as below shown:
image
This cmd is actually defined but not being called in Zephyr bt_tester anywhere, we want to enable this cmd and fill it with `bt_disable `to  achieve power off controller purpose
image

**Describe alternatives you've considered**

**Additional context**
I'm working on a PR to enable and fill this command, will submit it soon, please help to review and suggest, thanks in advance!!

### PR DESCRIPTION
**Pull request by zhaynxp**
Fill BTP_GAP_SET_POWERED command to support power on/off BLE controller independently

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/67346

Signed-off-by: Ying Zhang <ying.zhang_2@nxp.com>

### CODE REVIEW

**Comment by zhaynxp** [comment]
The issue this PR solves is https://github.com/zephyrproject-rtos/zephyr/issues/67346 , please help to review, thanks!

**Comment by Thalley** [comment]
@zhaynxp you can link a PR to an issue using one of the GH keywords in your PR description: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue

**Comment by zhaynxp** [comment]
@Thalley , thanks for your comments, have linked issue to this PR. 
BTW, there is compile error in this PR workflow checking, said undeclared function 'bt_conn_cb_unregister', actually this function was created by my nxp colleague in PR #64599 , the feature developed in this PR was based on fixes in #64599, otherwise BLE connection can't be established after `bt_disable `--> `bt_enable` , 
so need below PR checked in first, could you please help to review both PR, and let us know if any action needed? 
https://github.com/zephyrproject-rtos/zephyr/pull/64599 
The `bt_disable `--> `bt_enable `--> `bt_disable `repeat test is mandatory to us no matter with bt-shell or bt-tester, and currently block our testing, so please help to review and merge these PRs ASAP. Thanks advance!

**Comment by Thalley** [comment]
> @Thalley , thanks for your comments, have linked issue to this PR.

Hi @zhaynxp - It is not linked correctly. You have to add something like `closes https://github.com/zephyrproject-rtos/zephyr/issues/67346` or `fixes https://github.com/zephyrproject-rtos/zephyr/issues/67346` to properly link it

**Comment by sjanc** [comment]
Hi,

I don't think zephyr should implement this command, it is meant for systems that can disable controller only (ie. Linux) without shutting down whole host/apps stack.   And you also disable already registered services on power off.

If you need to disable bluetooth in zephyr I think correct way would be to implement service unregister functions and than BTP user can unregister services (including GAP which could disable bluetooth subsystem).  Just note that unregister functions would have to properly cleanup things as current implementation only have stubs for services unregister (as we don't really use it in autopts for now)

**Comment by zhaynxp** [comment]
> > @Thalley , thanks for your comments, have linked issue to this PR.
> 
> Hi @zhaynxp - It is not linked correctly. You have to add something like `closes https://github.com/zephyrproject-rtos/zephyr/issues/67346` or `fixes https://github.com/zephyrproject-rtos/zephyr/issues/67346` to properly link it

Hi @Thalley , I've add fixes #

**Comment by zhaynxp** [comment]
> > @Thalley , thanks for your comments, have linked issue to this PR.
> 
> Hi @zhaynxp - It is not linked correctly. You have to add something like `closes https://github.com/zephyrproject-rtos/zephyr/issues/67346` or `fixes https://github.com/zephyrproject-rtos/zephyr/issues/67346` to properly link it

Hi @Thalley , thanks! I've submitted new commit message to link issue, please help check and review. Thanks!

**Comment by zhaynxp** [comment]
> Hi,
> 
> I don't think zephyr should implement this command, it is meant for systems that can disable controller only (ie. Linux) without shutting down whole host/apps stack. And you also disable already registered services on power off.
> 
> If you need to disable bluetooth in zephyr I think correct way would be to implement service unregister functions and than BTP user can unregister services (including GAP which could disable bluetooth subsystem). Just note that unregister functions would have to properly cleanup things as current implementation only have stubs for services unregister (as we don't really use it in autopts for now)

Hi @sjanc , thanks a lot for your suggestion, we'd deprecated the BTP_GAP_SET_POWERED command, and use Unregister GAP Service command(00 04 ff 01 00 01) instead to shutdown whole Bluetooth Subsystem, it will act as **main switch** to power off BT controller and clean up app+host stack layer resources, and will call `unregister_all_services_except_gap` to unregister all other services such as GATT, L2CAP.. as you know, user/customer may registered lots of services during use, when they want to shutdown Bluetooth, it's hard/impossible to unregister the services one by one manually, so a main switch is convenient and  important we think, and we made it all cleaned with one Unregister GAP Service command. Please help to review the new solution and commit, thank you very much!

**Comment by zhaynxp** [comment]
Dear Reviewers,

Since this PR is on top of PR #64599 which completed bsim test recently and ready to merge soon if no accident, want to check  what's your opinion on this PR? Is it ok to merge after PR #64599 merged and all checks passed? We have a tight schedule to upstream these PRs, so please kindly share your views and let us know any pending items on our side, thanks a lot!

File: tests/bluetooth/tester/src/btp_gatt.c
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_gatt.c:2614]:
```suggestion
	if (server_buf) {
		net_buf_destroy(server_buf);
	}
	
	return BTP_STATUS_SUCCESS;
```

**Review by Thalley — CHANGES_REQUESTED** [review]
Some comments, but ideally @sjanc is the one that really needs to look at this

File: tests/bluetooth/tester/src/btp_gap.c
Comment (zhaynxp) [reply tests/bluetooth/tester/src/btp_gap.c:1884]: Hi Thalley, thanks a lot for your comments and suggestions, will correct above.
back to this question, requirement is from our QA that need an interface to shutdown Bluetooth independently during COEX running(we have coex_bt_tester app), to ensure Bluetooth and other radio will not affect each other. 
the bt_tester is not only used for autopts test, but also help us do some sanity tests, so have such request. 
Thanks!

Comment (zhaynxp) [reply tests/bluetooth/tester/src/btp_gap.c:1884]: also previouly we preferred to use BTP_GAP_SET_POWERED command to shutdown Bluetooth, then changed to Unregister GAP Service command(00 04 ff 01 00 01) per @sjanc 's suggestion.

Comment (Thalley) [reply tests/bluetooth/tester/src/btp_gap.c:1884]: I see - Thanks for the explanation. Let's hear what @sjanc has to say :)

**Comment by zhaynxp** [comment]
Dear Reviewers,

Would you please help to review solution in this PR? Since currently no unregister services will be called by autopts or other vendor, also obviously bt_tester lacks such an interface to call `bt_disable` , so guess this should be fine. Kindly help to review it again, many thanks!

**Comment by Thalley** [comment]
> Dear Reviewers,
> 
> Would you please help to review solution in this PR? Since currently no unregister services will be called by autopts or other vendor, also obviously bt_tester lacks such an interface to call `bt_disable` , so guess this should be fine. Kindly help to review it again, many thanks!

Please be patient :) I don't have any significant change requests, but @sjanc is effectively the maintainer of the BT tester, so I'd like to wait for him to provide a review.

**Comment by zhaynxp** [comment]
> Hi,
> 
> sorry for late response, BTP protocol doesn't allow for implicit unregistering of services
> 
> Since Zephyr's host doesn't really allow for shutdown, but rather full re-init, my suggestion would be to implement this per service. As GAP profile is kinda special it may for example return error as long as any other service is stil registered. And than that specifics can be handled in test system that is using tester application.
> 
> Other option could be extending BTP to allow for this behavior (eg a command in core service) but I don't have clear idea on how such command should look like (probably should allow for some customization to accommodate for various stacks)

Hi @sjanc , thanks for your reply. so you mean should not unregister other services inside unregister GAP service? and it should be the test system or user who use bt_tester themselves to unregister the registered services one by one, right? 
Also is it ok to call `bt_disable` inside unregister GAP service? as introduced in issue #67346 , we need an interface to call the `bt_disable` to shutdown Bluetooth subsystem, similar as the `bt disable` cmd in bt_shell app. 
Could you please help to double confirm? Thanks a lot!

**Comment by zhaynxp** [comment]
Hi @sjanc , 

As introduced in #67346 , our demand is to have an interface to call `bt_disable` in bt_tester so as to power off BLE controller during COEX test, and can work normally after power on the controller again. The reason why we clear app resource and unregister services in this PR was because there exists unable to connect issue with bt_tester after power on controller again. 
As you know, this topic is pending for a long time, we reviewed this case and have few possible proposals as below, could you please help  to review and share your opinion ASAP? Thanks a lot！
a.  use below BTP_GAP_RESET command to call `bt_disable` + `bt_enable `to achieve restart BLE controller feature
image
     advantages of this way is : reset controller with only 1 cmd + no unregister services and clear app resource needed, the already registered services and command handlers can keep running
b. use below BTP_GAP_SET_POWERED command to power on/off controller, if the cmd parameter Powered is 1 then call `bt_enable `to power on controller, else call `bt_disable `to power off controller
image
     advantages of this way is : no unregister services and clear app resource needed+ can disable/enable controller separately
c. as submitted in this PR, or the comment last week, use unregister GAP service to call `bt_disable `
d. no way to call `bt_disable`, won't support this, and share the reason
Could you please let us know your preference explicitly so we can proceed this ticket accordingly? Thanks again!

**Comment by sjanc** [comment]
Hi,

Okay, so I thought about this, double checked APIs and discussed this with with Emil.
Conclusion was that it should be possible to implement 'Set powered' command by simply mapping this to bt_disable() and bt_enable() API.

It is not ideal but best we can get regarding BTP and what Zephyr offers. Just make sure to properly toggle BTP_GAP_SETTINGS_POWERED BTP setting when doing so (via BTP_GAP_EV_NEW_SETTINGS event)

**Comment by Thalley** [comment]
> Hi,
> 
> Okay, so I thought about this, double checked APIs and discussed this with with Emil. Conclusion was that it should be possible to implement 'Set powered' command by simply mapping this to bt_disable() and bt_enable() API.
> 
> It is not ideal but best we can get regarding BTP and what Zephyr offers. Just make sure to properly toggle BTP_GAP_SETTINGS_POWERED BTP setting when doing so (via BTP_GAP_EV_NEW_SETTINGS event)

It is, however, very important to note that bt_disable() followed by a bt_enable() does not mimic a reboot. At best it mimics this in the controller, but there are data in the host that is not cleared (most notably registered services, callbacks, etc. (anything that can be changed before calling bt_enable())).

**Comment by zhaynxp** [comment]
Hi @sjanc @Thalley ,

Thank you very much for the confirmation, agree that currently the set powered command seems to be the most appropriate for the power on/off controller scenario. Well understand it's not system/BT reboot, but can help controller reboot without touch app layer resources. Have submitted the PR for review, kindly pls help review and let us know any comments. 
Many thanks!

**Review by sjanc — COMMENTED** [review]
enable set powered command bit in supported_commands(),  otherwise looks OK to me

**Comment by zhaynxp** [comment]
> enable set powered command bit in supported_commands(), otherwise looks OK to me
 
thanks sjanc for the reminder ! Have corrected this

**Review by sjanc — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
